Add MockL2OutputOracle contract and L2Output Submitter

cfromknecht opened the pull request:
**Description**
This PR adds a working L2Output batch submitter and integrates it with E2E test harness. To achieve this, a stripped down copy of the EIP-1559 batch submitter is added, and the `l2output` driver is implemented which interacts with ABI bindings for an initial iteration of the `MockL2OutputOracle.sol` contract. The first 5 or so commits could probably be squashed down, though I have left them separate for now to make it clear what was removed from the existing batch submitter. ~There is still some trimming of config options to be done, which I will triage at a later point after getting some feedback on the initial approach.~ For now the batch submitter has it's own `bss` binary, but this can be modified if we prefer a unified binary.

**Metadata**
- Fixes ENG-1928

trianglesphere commented:
One large organization question (I only skimmed, didn't dive deep), do we intend to have the `bss` be a separate binary from the `opnode`. It feels like it's exactly copying the structure from the monorepo, but this feels a little odd to have this as a subproject while the `opnode` is a top level project.

cfromknecht commented:
@trianglesphere I don't have a super strong preference on whether the binaries are separate or the batch submitter is a subcommand. I would lean towards in production running them as separate services in either case though. For now i just left it as a separate binary, since it doesn't impact our ability to integrate with the E2E tests. If there's a strong preference for merging everything into the `opnode` command immediately we can make that happen.

tynes commented:
I definitely prefer to run them as different binaries so that they hot key can be managed more easily

trianglesphere commented on opnode/bss/Makefile:
```diff
@@ -0,0 +1,32 @@
+LDFLAGSSTRING +=-X main.GitVersion=$(GITVERSION)
+LDFLAGS := -ldflags "$(LDFLAGSSTRING)"
+
+CTC_ABI_ARTIFACT := ../../packages/contracts/artifacts/contracts/L1/rollup/CanonicalTransactionChain.sol/CanonicalTransactionChain.json
```
Leftover? These otherwise don't appear to be checked in.

cfromknecht replied:
Yes good catch, removed.

trianglesphere commented on opnode/bss/config.go:
```diff
@@ -0,0 +1,224 @@
+var (
+	// ErrSequencerPrivKeyOrMnemonic signals that the user tried to set both
+	// sequencer wallet derivation methods or neither of them.
+	ErrSequencerPrivKeyOrMnemonic = errors.New("either sequencer-private-key " +
```
Splitting the line helps with line length, but makes it hard to grep for the exact error from the error string.

cfromknecht replied:
This error has been removed, but i did a quick grep and it seems like all of the other instances are on one line 👍

trianglesphere commented on opnode/test/system_test.go:
```diff
@@ -125,10 +128,57 @@ func TestSystemE2E(t *testing.T) {
+	require.Nil(t, err)
+
+	opts, err := bind.NewKeyedTransactorWithChainID(
+		l2OutputPrivKey, big.NewInt(900),
```
Could be pulled from the config

cfromknecht replied:
done

trianglesphere commented on opnode/test/system_test.go:
```diff
@@ -125,10 +128,57 @@ func TestSystemE2E(t *testing.T) {
+	batchSubmitter, err := bss.NewBatchSubmitter(bss.Config{
+		BuildEnv:               "develop",
+		EthNetworkName:         "clique",
+		L1EthRpc:               httpHost(cfg.l1.nodeConfig.HTTPHost, cfg.l1.nodeConfig.HTTPPort),
```
`endpoint(cfg.l1.nodeConfig)` returns the WS endpoint. It could be renamed to `wsEndpoint` and a `httpEndpoint(...)` could be added if the batch submitter prefers using HTTP

cfromknecht replied:
i hadn't thought to try the bss with WS's, but it seems to work just fine. removed the `httpHost` logic

trianglesphere commented on opnode/bss/package.json:
```diff
@@ -0,0 +1,6 @@
+{
+	"name": "@eth-optimism/batch-submitter-service",
```
Do we have docs on this/the changelog?

cfromknecht replied:
I've removed the changelog for now, we can add it at a later time if we want

trianglesphere reviewed (commented):
Thanks for clarifying that the bss will be it's own binary/service. With that it feels like the directory structure has it as a submodule rather than a peer to the deposit sequencer node.

I know this was copied from the mono repo, but I wonder if the bss needs it's own `go.mod` file - i.e. will it be versioned independently from the rest of the rollup node components.

tynes commented on opnode/bss/config.go:
```diff
@@ -0,0 +1,224 @@
+	// Mnemonic is the HD seed used to derive the wallet private keys for both
+	// the sequence and l2output. Must be used in conjunction with
+	// SequencerHDPath and L2OutputHDPath.
+	Mnemonic string
+
+	// SequencerHDPath is the derivation path used to obtain the private key for
+	// the sequencer transactions.
+	SequencerHDPath string
+
+	// L2OutputHDPath is the derivation path used to obtain the private key for
+	// the l2output transactions.
+	L2OutputHDPath string
```
This is how we configure the existing batch submitter. Using an encrypted keystore (https://github.com/roynalnaruto/eth-keystore-rs) could help to improve our security practices

tynes replied:
I wish that encrypted keystores worked with mnemonics by holding information about the hd path as part of its datastructure. Or if the encrypted keystore could hold a mnemonic and then the app could choose to derive whatever key it wants

cfromknecht replied:
> Why do we import a mnemonic and specify derivation paths? I think we should instead specify a file-path towards an encrypted keystore and another to a keystore password, and mount/protect that password file with some key vault when deployed in production. IMHO mnemonics should only ever be used outside of deployed code, for backups etc. and never for key inputs.

I don't know the specifics of how it's implemented but in prod we use Google's secret manager to inject these. I have to imagine under the hood the setup is fairly similar to what you describe here—the mnemonic is not stored in a config file if that's your concern. As far using a raw key vs mnemonic+hdpath I don't have a strong preference. In the current BSS we use the mnemonic since it allows us to derive a key for each service (tx and state-roots) using only 1 pice of secret data, rather than multiple.

tynes commented on opnode/bss/crypto.go:
```diff
@@ -0,0 +1,108 @@
+	"fmt"
+	"strings"
+
+	"github.com/decred/dcrd/hdkeychain/v3"
```
Perhaps https://github.com/miguelmota/go-ethereum-hdwallet ?

cfromknecht replied:
`go-ethereum-hdwallet` is just a wrapper around `btcutil/hdkeychain` for the things we use it for, but swapping it out did allow us to get rid of the `crypto.go` and `crypto_test.go` files 👍

tynes commented on opnode/bss/config.go:
```diff
@@ -0,0 +1,224 @@
+
+	// PropopserPrivateKey the private key of the wallet used to submit
+	// transaction to the SCC contract.
+	L2OutputPrivateKey string
```
These config options were used at one point with the old batch submitter

cfromknecht replied:
yes this config option is used in various test environments in the monorepo. i went ahead and removed the option to set private keys explicitly since this test environment already uses mnemonics

cfromknecht commented on opnode/bss/CHANGELOG.md:
```diff
@@ -0,0 +1,17 @@
+# @eth-optimism/batch-submitter-service
```
removed the changelog, we can add it if we decide we want two separate services

cfromknecht commented:
@protolambda @trianglesphere thanks for the reviews! I was able to whittle down the remainder of the config options and code down to a more minimal set.

The main thing that comes to mind rn—are we okay with leaving the `bss` binary separate? Or should I go ahead and make it a subcommand? The main thing here is that we currently use two different libraries for config parsing (`ask` for `opnode`, and `urfave/cli` for `bss`). I'd like to solidify on around one before proceeding.

@trianglesphere I moved the batch submitter from `opnode/bss` to `bss` given your prior comment, and also removed the `bss/go.mod` so that it is part of the top-level module.

trianglesphere commented:
> The main thing that comes to mind rn—are we okay with leaving the `bss` binary separate? Or should I go ahead and make it a subcommand? The main thing here is that we currently use two different libraries for config parsing (`ask` for `opnode`, and `urfave/cli` for `bss`). I'd like to solidify around one before proceeding.

I think separate binaries are fine for now. Given that they are planned to be ran separately I don't see a large benefit in unifying them under one command.

trianglesphere commented:
Completely separately (and not something that should be solved in this PR), have we considered making the tx manager a separate service (like it's own process) that we then connect to (probably need some sort of authentication).

Another idea here is to pull out the tx manager to a shared package - it seems like a good candidate given how much we use it from different contexts.

protolambda reviewed (approved):
Over time I definitely think we should use encrypted keystores as opposed to mnemonics (best security practice, don't want a leak to leak all other account private keys derivable from a mnemonic), but otherwise this all looks good.

norswap reviewed (approved):
Interesting stuff!

Made some minor comments (typos, etc), and asked a few questions where the logic is not entirely clear to me + some high level items.

Would be interested to know what the future work on this is. Needs an issue to track it if this is merged.

norswap commented:
Related to #170 (epic: #156)

maurelian commented on packages/contracts/contracts/L1/StateRootOracle.sol:
```diff
@@ -0,0 +1,99 @@
+        return (latestBlockTimestamp - startingBlockTimestamp) / submissionFrequency;
+    }
+
+    function deleteStateBatch(uint _timestamp) external {
```
```suggestion
    function deleteState(uint _timestamp) external {
```

cfromknecht replied:
removed this method 👍

maurelian commented on packages/contracts/contracts/L1/StateRootOracle.sol:
```diff
@@ -0,0 +1,99 @@
+// We will eventually need to add support for concurrent
+// state root proposals, but until we have a full dispute
+// game spec that we are happy with, it is better to KEEP IT SIMPLE.
+contract StateRootOracle {
```
@cfromknecht for the purpose of this PR, would it be possible to add just the interface for this contract, or the minimum viable amount of contract you need to implement the go-code?

cfromknecht replied:
No problem, I stripped down the contract to get to an MVP 👍

maurelian reviewed (changes requested):
I'd prefer not to merge a new contract without testing. If necessary for the sake of implementing and testing the go code, we could create a Mock instead.

cfromknecht commented on bss/flags/flags.go:
```diff
@@ -0,0 +1,105 @@
+	PollIntervalFlag = cli.DurationFlag{
+		Name: "poll-interval",
+		Usage: "Delay between querying L2 for more transactions and " +
+			"creating a new batch",
```
the `cli.DurationFlag` parses inputs of the form `1s`, `500ms`, `2h`, there is no inherent unit of time

norswap replied:
Is that a common Go thing? Otherwise might be worth mentionning that you can pass the unit. What if you don't mention a unit?

cfromknecht replied:
yes it’s pretty common for go. under the hood urfave just calls into the stdlib `time.ParseDuration` https://pkg.go.dev/time#ParseDuration

if no units are provided it defaults to ns

cfromknecht commented on bss/txmgr/txmgr.go:
```diff
@@ -0,0 +1,274 @@
+// that error indicates that none of the transactions will ever confirm.
+func shouldAbortImmediately(err error) bool {
+	return strings.Contains(err.Error(), core.ErrNonceTooLow.Error())
+}
```
A tx sent by a prior running instance confirming is the most likely.

> Also, can't this be handled by bumping up the nonce and retrying?

Yes, that's exactly what happens. The submission at the current nonce is aborted, and the system reevaluates what needs to be done at the next nonce and if necessary calls back into `Send`. We don't want to blindly bump the nonce here as the confirming txn may contain the same data.

cfromknecht commented on bss/txmgr/txmgr.go:
```diff
@@ -0,0 +1,274 @@
+			tipHeight, err := backend.BlockNumber(ctx)
+			if err != nil {
+				log.Error("Unable to fetch block number", "err", err)
+				break
```
Correct. If this fails while reorging we will just sleep and wait to try again

cfromknecht commented on bss/Makefile:
```diff
@@ -0,0 +1,26 @@
+GITCOMMIT := $(shell git rev-parse HEAD)
+GITDATE := $(shell git show -s --format='%ct')
+GITVERSION := $(shell echo 0.0.0)
```
The `GITVERSION` name is a little confusing. This should actually just be `VERSION`. In prod this is pulled from the `package.json` which is updated by `changeset`, but afaict we don't have that set up in this repo so I just made a placeholder.

norswap replied:
Could you make a comment to that effect?

cfromknecht commented on bss/flags/flags.go:
```diff
@@ -0,0 +1,105 @@
+	MaxBatchSubmissionTimeFlag = cli.DurationFlag{
+		Name: "max-batch-submission-time",
+		Usage: "Maximum amount of time that we will wait before " +
+			"submitting an under-sized batch",
```
turns out we don't use this flag, removed

cfromknecht commented on bss/config.go:
```diff
@@ -0,0 +1,87 @@
+	L2EthRpc string
+
+	// SROAddress is the SCC contract address.
+	SROAddress string
```
oops, the `SCC` part was left over from when this was renamed from `SCCAddress`. i've replaced `SCC` with `SRO` since this matches the current name of the solidity file, but we can definitely change it when we agree on a new name :)

cfromknecht commented on bss/config.go:
```diff
@@ -0,0 +1,87 @@
+// ensure that it is well-formed.
+func ValidateConfig(cfg *Config) error {
+	// Sanity check log level.
+	_, err := log.LvlFromString(cfg.LogLevel)
```
I ended up removing `ValidateConfig` altogether, it's really not as useful here as it is in the production BSS where we do some other sanity checks.

cfromknecht commented on bss/batch_submitter.go:
```diff
@@ -0,0 +1,178 @@
+
+		log.Info("Batch submitter started")
+
+		<-(chan struct{})(nil)
```
yes this will block forever. it's not super critical to add signal handling since we don't have really have anything that needs proper closing, i.e. a database connection, but added it anyways just in case 👍

cfromknecht commented on bss/batch_submitter.go:
```diff
@@ -0,0 +1,178 @@
+		return nil, err
+	}
+
+	l2Client, err := dialL1EthClientWithTimeout(ctx, cfg.L2EthRpc)
```
done!

cfromknecht commented on bss/drivers/l2output/driver.go:
```diff
@@ -0,0 +1,230 @@
+	}
+
+	parsed, err := abi.JSON(strings.NewReader(
+		sro.StateRootOracleABI,
```
fixed

cfromknecht commented on bss/drivers/l2output/driver.go:
```diff
@@ -0,0 +1,230 @@
+	if nextCheckpointBlock.Cmp(expCheckpointBlock) != 0 {
+		panic(fmt.Sprintf("next expected checkpoint block has changed, "+
+			"want: %d, found: %d", nextCheckpointBlock.Uint64(),
+			expCheckpointBlock.Uint64()))
```
I don't think so, if we're executing this logic that means the prior submission happened 25 minutes/100 blocks in the past, which would be a very deep reorg. I changed it to return an error just in case tho, which would be safer.

cfromknecht commented on bss/drivers/l2output/driver.go:
```diff
@@ -0,0 +1,230 @@
+
+	numElements := new(big.Int).Sub(start, end).Uint64()
+	log.Info(name+" batch constructed", "num_state_roots", numElements)
+
```
I worked the logging slightly here to call it a checkpoint rather than a batch, which I think works well. Happy to change the terminology further

norswap replied:
So a checkpoint will be batched transactions + output commitment, with the transaction part not yet implemented, is that the idea?

cfromknecht commented on packages/contracts/contracts/L1/StateRootOracle.sol:
```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: MIT
```
> This file has mixed tabs and spaces, and has a lot of indentation / formatting issue. I'll make a commit to fix those (feel free to revert it if you made concurrent changes to this file).

Realized this was an outdated version of the contract, should be better now.

> Also we probably should rename "state root" into "l2 output commitment".

So `L2OutputCommitmentOracle`? Could also be something like `L2OutputCheckpointOracle` or`L2CheckpointOracle`

norswap replied:
All of these are fine by me, I guess "output commitment" is what is predominantly used in the code? (I don't remember!)

cfromknecht commented:
> I'd prefer not to merge a new contract without testing. If necessary for the sake of implementing and testing the go code, we could create a Mock instead.

@maurelian I've never written solidity tests before but I can give it a shot. The contract has some coverage through the e2e test, which might be an easier place to expand for the sake of this PR

maurelian commented on bss/bindings/sro/state_root_oracle.go:
```diff
@@ -0,0 +1,616 @@
+// Code generated - DO NOT EDIT.
```
Might be a naive Go question, but should this be checked-in, or generated by the build?

trianglesphere replied:
Generally we'd like to have these checked in.

protolambda replied:
Generated Go code should be checked in. We can use the `go:generate` thing to automate re-generating it, and we should add something to CI to re-generate everything and diff with the existing files, to make sure the generated code is correct.

trianglesphere replied:
FYI a lot of these are generated with `make` (easier to handle with the output of `hardhat`), but adding a CI step to check that the generated files are up to date would be a nice to have.

maurelian commented:
I too hate when someone just drives by, leaves a few comments, then requests changes without clearly stating what it would take to get their approval. So I'll clarify/enumerate what I'd like to see prior to approving this PR: 

1. [ ] Rename the contract by prefixing it with 'Mock' and run 'yarn lint' on it.  
  - It's untested and full of holes (ie. no auth on `deleteStatebatch`). I'd prefer to even reduce the functionality to a mock, but a renaming to make it clear we're not merging broken reference code would be OK with me. 
2. [ ] Update names from the "Batch Submission" paradigm to match with the "L2 Output Submission"  paradigm.

I'd also _prefer_ to have specs in before ref impl, but it's kind of a chicken and egg thing, so I don't want to be overly pedantic. That said, it just means that when the spec (#194) comes in, it will need to be reviewed to ensure it matches the impl.

cfromknecht commented:
> Rename the contract by prefixing it with 'Mock' and run 'yarn lint' on it.

Done!

> Update names from the "Batch Submission" paradigm to match with the "L2 Output Submission" paradigm.

@maurelian I've renamed the top level `BatchSubmitter` struct to `L2OutputSubmitter`, updated log lines, help msg, etc., and removed all reference of batches in comments favor of L2Outputs. @norswap I've renamed the package `l2os` for L2 Output Submitter, but if you have a better name I can change it again.

I do wonder how future proof naming the top-level struct `L2OutputSubmitter` is—once we add logic in to submit tx batches it won't be quite accurate. I wonder if there's a better name that would be more generic, but we can also deal with that bridge when we get there as it is accurate for the time being.

cfromknecht commented:
I'd like to suggest holding off on merging until https://github.com/ethereum-optimism/optimism/pull/2165 lands, that way I can update this PR with the fix. Also another reason to make a standalone `txmgr` package..

maurelian commented:
> I'd like to suggest holding off on merging until [ethereum-optimism/optimism#2165](https://github.com/ethereum-optimism/optimism/pull/2165) lands, that way I can update this PR with the fix. Also another reason to make a standalone `txmgr` package..

OK, setting to draft.

trianglesphere commented on l2os/cmd/l2os/main.go:
```diff
@@ -0,0 +1,44 @@
+
+	app := cli.NewApp()
+	app.Flags = flags.Flags
+	app.Version = fmt.Sprintf("%s-%s", Version, params.VersionWithCommit(GitCommit, GitDate))
```
FYI this pulls from geth's hardcoded version.

trianglesphere replied:
yep

cfromknecht replied:
Fixed, just used a simple `<version>-<commit>-<date>` format

maurelian commented:
@cfromknecht I see that https://github.com/ethereum-optimism/optimism/pull/2165 has been merged. 
What does that mean for the status of this PR? Does it need to incorporate changes from there?

mslipper commented:
I believe this is ready for merge.

cfromknecht commented:
> I believe this is ready for merge.

Rebased!

maurelian commented:
I'm hesitant to merge this large PR given the gaps in our review process we've recently discussed. So I've asked @trianglesphere to give it another review, but first I'd appreciate if @cfromknecht could add the following in the way of details:

1. What are the security risks in this code and how are they mitigated? 
2. Which areas of the code require the most scrutiny?

trianglesphere reviewed (approved):
I think we should go ahead and merge this.

Looking through the code, I was looking for specs, Frequency/the exact output, and internal states that can get messed up.

My impression is that we need to follow up with specs on how often we should commit to the output. This is based on my experience, but I think we rely too much on timestamps when determining when we should have an output commitment.

Some other thoughts on specs (more from the side of verifying the output commitment)
- Over what range is the commitment
- How delayed can the commitment be from when it is available to be committed to and when it must be on chain (I guess there is an economic staking component to this) (or can anyone submit an update after a while)
- How do withdrawals interact with having outputs (and not having them)
- Do we want to tie output commitments to a specific chain (i.e. include a parameter for the parent block hash (if this can be done))

trianglesphere commented:
@cfromknecht 🚢 it

JahJah01 commented:
Thax for the merge

JahJah01 commented:
Thax for the merge it was a grade experience

JahJah01 commented:
Thax merge

JahJah01 commented:
Thak you
